### PR TITLE
(Step 4/5) Add KPLs to RetinaNet training script

### PR DESCRIPTION
**Pull request by LukeWood**
Notes:

- The checked in version of this script doesn't work at all
- This scores the same MaP as before, >0.32 every run
- This is much more readable
- This adds back in the COCO metrics, these are really useful for monitoring training on a GPU device.  We can disable them on TPU if we want, but I'd like to keep them for prototyping until we hit our final state
- I've removed the COCO evaluator callback until we move everything to native box format

---

Action items after this PR:

- [ ] swap to native box format
- [ ] add back the COCO callback
- [ ] fix the label encoder metric in MirroredStrategy mode

### CODE REVIEW

**Review by tanzhenyu — COMMENTED** [review]
Interesting...I think COCOMetrics doesn't work on GPU as well (What's unique about TPU here?)

File: examples/training/object_detection/pascal_voc/retina_net.py [hunk 1]
```diff
@@ -363,31 +170,33 @@ def pad_fn(image, boxes):
...
     keras.callbacks.ReduceLROnPlateau(patience=5),
     keras.callbacks.EarlyStopping(patience=10),
     keras.callbacks.ModelCheckpoint(CHECKPOINT_PATH, save_weights_only=True),
-    PyCOCOCallback(eval_ds.map(convert_to_dict), "xywh"),
 ]
 
 history = model.fit(
     train_ds,
-    epochs=100,
+    validation_data=eval_ds.take(20),
```
Comment (ianstenbit) [inline examples/training/object_detection/pascal_voc/retina_net.py:199]: Is the call to `.take` here intentional? I'd think we would want the whole eval DS?

**Comment by ianstenbit** [comment]
> Interesting...I think COCOMetrics doesn't work on GPU as well (What's unique about TPU here?)

IIRC the root cause of COCOMetrics not working on TPU was the NMS issue. So in that case I'd expect it to work fine on GPU

**Review by ianstenbit — APPROVED** [review]
Looks a lot nicer!

**Comment by bhack** [comment]
Wasn't it TPU only cause we cannot XLA compile for https://github.com/tensorflow/tensorflow/issues/56769#issuecomment-1184793361?

**Comment by bhack** [comment]
Last time was https://github.com/keras-team/keras-cv/issues/895#issuecomment-1273430678

If was NMS it is TPU compatible with mlir bridge as we have tested in another ticket.

**Comment by bhack** [comment]
The MMS test was at https://github.com/keras-team/keras-cv/issues/895#issuecomment-1273430678

**Comment by tanzhenyu** [comment]
> > Interesting...I think COCOMetrics doesn't work on GPU as well (What's unique about TPU here?)
> 
> IIRC the root cause of COCOMetrics not working on TPU was the NMS issue. So in that case I'd expect it to work fine on GPU

That's a graph building time error. I mean realistically can TensorArray even work on GPU?

**Comment by tanzhenyu** [comment]
> > > > Interesting...I think COCOMetrics doesn't work on GPU as well (What's unique about TPU here?)
> > > 
> > > 
> > > IIRC the root cause of COCOMetrics not working on TPU was the NMS issue. So in that case I'd expect it to work fine on GPU
> > 
> > 
> > That's a graph building time error. I mean realistically can TensorArray even work on GPU?
> 
> yeah, COCO metrics work on GPU. the only thing broken on GPU is the label encoder's % boxes caputured encoding tracker

I see. I haven't closely follow the implementation. Shared a colab link with you, it seems it doesn't work when we force everything to be placed on GPU.

**Comment by LukeWood** [comment]
Lets keep this lightweight for now, I am mainly using this as a way to ensure we still get SOTA MaP.  When I migrate to dictionary format across all models I'll migrate this to use the callback.

**Comment by bhack** [comment]
> Shared a colab link with you, it seems it doesn't work when we force everything to be placed on GPU.

Can you share the colab?
